src/tests/unit/common/utils/test_resolve_reference.py

```diff
@@ -0,0 +1,169 @@
+                "type": SIMOS.REFERENCE.value,
+                "referenceType": REFERENCE_TYPES.LINK.value,
+            },
+        }
+        self.engine = {
+            "_id": "2",
+            "type": "test_data/complex/EngineTest",
+            "name": "myEngine",
+            "description": "",
+            "fuelPump": {
+                "address": "$3",
+                "type": SIMOS.REFERENCE.value,
+                "referenceType": REFERENCE_TYPES.LINK.value,
+            },
+            "power": 120,
+        }
+        self.fuel_pump = {
+            "_id": "3",
+            "type": "test_data/complex/FuelPumpTest",
+            "name": "fuelPump",
+            "description": "A standard fuel pump",
+        }
+        self.document_repository = mock.Mock()
+        self.document_repository.name = "datasource"
+        self.document_repository.get = self.mock_get
+        self.document_repository.find = self.mock_find
+        self.document_service = get_mock_document_service(lambda x, y: self.document_repository)
+
+    def mock_get(self, document_id: str):
+        if document_id == "1":
+            return {**self.car_rental_company}
+        if document_id == "2":
+            return {**self.engine}
+        if document_id == "3":
+            return {**self.fuel_pump}
+        if document_id == "4":
+            return {**self.customer}
+        return None
+
+    def mock_find(self, query: dict) -> list[dict]:
+        documents: list[dict] = [self.car_rental_company, self.engine, self.fuel_pump]
+        for key, value in query.items():
+            documents = list(filter(lambda x: key in x and x[key] == value, documents))
+        return documents
+
+    def test_id(self):
+        ref = resolve_reference("datasource/$1", self.document_service.get_data_source)
+        assert ref.data_source_id == "datasource"
+        assert ref.document_id == "1"
+        assert ref.attribute_path == []
+        assert ref.entity == self.car_rental_company
+
+    def test_with_attributes(self):
+        ref = resolve_reference("datasource/$1.cars[0]", self.document_service.get_data_source)
+        assert ref.data_source_id == "datasource"
+        assert ref.document_id == "1"
+        assert ref.attribute_path == ["cars", "[0]"]
+        assert ref.entity == self.car_rental_company["cars"][0]
+
+    def test_with_attributes_to_uncontained(self):
+        ref = resolve_reference("datasource/$1.cars[0].engine", self.document_service.get_data_source)
+        assert ref.data_source_id == "datasource"
+        assert ref.document_id == "1"
+        assert ref.attribute_path == ["cars", "[0]", "engine"]
+        assert ref.entity == self.car_rental_company["cars"][0]["engine"]
+
+    def test_with_attributes_to_uncontained_child(self):
+        ref = resolve_reference("datasource/$1.cars[0].engine.fuelPump", self.document_service.get_data_source)
+        assert ref.data_source_id == "datasource"
+        assert ref.document_id == "2"
```

Comment:
The nearest parent I think. resolve_reference does not resolve the last entity, since that would make it difficult to change or delete references. So, since ref.entity is a reference located in the self.engine document, it makes sense that the document_id is the id of self.engine.